Refactor I/O

Yurlungur opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

In a step towards getting the node/face/edge centered data to work, I am moving some of the generic I/O code out of HDF5 into utils. This should also help with issue #946 and getting ADIOS/IO to work. A main part of this effort is abstracting out those big gross nested for loops, which need to be done both at write time and read time, so that we can change things in only one place rather than have to synchronize them across files.

@bprather this is where I'm tweaking things.

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

jonahm-LANL committed:
allow for disabling of XDMF

jonahm-LANL committed:
Move output loops into PackOrUnpackVar

pgrete commented:
Let me know when you want comments here.

jonahm-LANL committed:
Merge branch 'develop' into jmm/io-refactor

c-prather commented:
Linking #752 here.  It's fine if you don't want to tackle that in this PR, but given leave I'd like to try addressing that here when you're done with the indexing improvements.

Basically I'd like to move all the logic about output types and file names upward in the stack, and make them all use a common `MakeOutput(vector<string> variable_list, string filename)` call.  Then filenames can be more flexible, and new outputs are easier to define in downstream codes, not to mention implementing all the features in that issue becomes trivial.

jonahm-LANL committed:
tiny cleanup with size method in VarInfo struct

jonahm-LANL committed:
move some logic into the mesh

jonahm-LANL committed:
move compute coords into I/O utils

jonahm-LANL committed:
Remove unneeded prototypes from PHDF5Output class

jonahm-LANL committed:
Mark TopologicalOffset constexpr so that we can simplify the coords logic

jonahm-LANL committed:
move WriteBlocksMetadata into a helper function

jonahm-LANL committed:
oops fix typo

jonahm-LANL committed:
locations into subfunction

Yurlungur commented on src/coordinates/uniform_cartesian.hpp:
```diff
@@ -164,18 +164,11 @@ class UniformCartesian {
-    [[maybe_unused]] bool constexpr X3EDGE =
-        el == TE::F3 || el == TE::E1 || el == TE::E2 || el == TE::NN;
-    if constexpr (dir == X1DIR && X1EDGE) {
+    if constexpr (dir == X1DIR && TopologicalOffsetI(el)) {
```
This is just miscellaneous cleanup. Not needed in this PR but also a small savings that we might as well do.

pgrete replied:
I think this makes it even clearer.

Yurlungur commented on src/mesh/mesh.cpp:
```diff
@@ -1204,6 +1204,22 @@ int Mesh::GetNumberOfMeshBlockCells() const {
 }
 const RegionSize &Mesh::GetBlockSize() const { return base_block_size; }
 
+const IndexShape &Mesh::GetLeafBlockCellBounds(CellLevel level) const {
+  // TODO(JMM): Luke this is for your Metadata::fine stuff.
+  PARTHENON_DEBUG_REQUIRE(level != CellLevel::fine,
+                          "Currently no access to finer cellbounds");
+  MeshBlock *pmb = block_list[0].get();
+  if (level == CellLevel::same) {
+    return pmb->cellbounds;
+    // TODO(JMM):
+    // } else if (level == CellLevel::fine) {
+    //   return pmb->fine_cellbounds;
+    // }
+  } else { // if (level == CellLevel::coarse) {
+    return pmb->c_cellbounds;
+  }
+}
```
cf @lroberts36

jonahm-LANL committed:
Move WriteLevelsAndLocs to subroutine

jonahm-LANL committed:
Fully disentangle the HDF5 shapes in the file

jonahm-LANL committed:
offset in 1 direction basically always 0

pgrete commented:
Is this ready for review/comments or is something (important) missing?
I'm asking as I'm working on the OpenPMD/ADIOS backend.

Yurlungur committed:
Merge branch 'develop' into jmm/io-refactor

Yurlungur commented:
It's not 100% all the changes I wanted to include, but we could review it and get it in as a first pass so you could leverage the changes I've made here.

pgrete commented on src/basic_types.hpp:
```diff
@@ -49,7 +49,7 @@ enum class TaskStatus { fail, complete, incomplete, iterate, skip, waiting };
 enum class AmrTag : int { derefine = -1, same = 0, refine = 1 };
 enum class RefinementOp_t { Prolongation, Restriction, None };
-
+enum class CellLevel : int { coarse = -1, same = 0, fine = 1 };
```
:+1:

pgrete commented on src/mesh/mesh.cpp:
```diff
@@ -1204,6 +1204,22 @@ int Mesh::GetNumberOfMeshBlockCells() const {
+  // TODO(JMM): Luke this is for your Metadata::fine stuff.
+  PARTHENON_DEBUG_REQUIRE(level != CellLevel::fine,
+                          "Currently no access to finer cellbounds");
+  MeshBlock *pmb = block_list[0].get();
```
What are the implicit assumptions/implications here always using the first block?

Yurlungur replied:
That all blocks have the same shape, which is actually not necessarily true in the multigrid case, but in that case there are separate block lists, so this one still is safe for that assumption.

Yurlungur replied:
Also that's why this is called `GetLeafBlockCellBounds`.

pgrete replied:
Gotcha, I blanked on the cellbounds just returning indices and nothing coordinate related.

pgrete commented on src/mesh/mesh.hpp:
```diff
@@ -194,6 +195,20 @@ class Mesh {
 
+  // TODO(JMM): Put in implementation file?
+  auto GetLevelsAndLogicalLocationsFlat() const noexcept {
+    std::vector<std::int64_t> levels, logicalLocations;
```
How about making the `logicalLocations` a vector of an array of size 3?
In memory (so to pass, e.g., to hdf5) it'd be identical, but in other places in the code it might be more readable.

Yurlungur replied:
In that case, is there any danger of, e.g., padding? With `std::vector<std::array<std::int64_t, 3>>`, I can imagine the compiler padding the end of the array struct... So I don't think it's actually guranteed to be identical in memory. I could make a vector of pointers to construct a C-style array? But that would be what's already here but with extra steps.

Alternatively, I could make something that pulls out the logical locations and then something else that flattens?

pgrete replied:
Why would there potentially be padding? And if there's padding this would not be impacted when passing the data pointer with the corresponding `.size()` call, does it?

Yurlungur replied:
Maybe I'm not understanding what you want to do---My understanding is that you want `std::vector<std::array<std::int64_t, 3>>` right? But is `std::array<T, N>` not a struct? The C++ standard specifies that structs can have arbitrary padding, so there may be extra bits between `&vec[0]` and `&vec[1]`, not just `3*64`. Or am I misunderstanding something?

pgrete commented on src/outputs/outputs.hpp:
```diff
@@ -64,12 +64,13 @@ struct OutputParameters {
         include_ghost_zones(false), cartesian_vector(false),
         single_precision_output(false), sparse_seed_nans(false),
-        hdf5_compression_level(5) {}
+        hdf5_compression_level(5), write_xdmf(true) {}
```
Any specific reason to use `true`as default?
hst, histogram, Ascent all do not need/support xdmf.

Yurlungur replied:
backwards compatibility only. If we're fine breaking that, I'm fine making this default to false.

Yurlungur replied:
Ah, I see what you mean. I've set the default initialized value to `false`, but the default value from `GetOrAddBoolean` when called by `ENABLE_HDF5` to `true`.

pgrete commented on src/parthenon_manager.cpp:
```diff
@@ -336,6 +336,9 @@ void ParthenonManager::RestartPackages(Mesh &rm, RestartReader &resfile) {
       // Double note that this also needs to be update in case
       // we update the HDF5 infrastructure!
+      // JMM: Version -1 is the old version. We are now on version 3.
+      // TODO(JMM): Do we need to maintain backwards compatibility
+      // here? It would be ncie to remove this.
```
I wouldn't mind to remove the old stuff (potentially in separate PR as it'd also pertain to the phdf reader) and mark is as deprecated.

Yurlungur replied:
:+1: I don't see a way to add the `[[deprecated]]` atribute to arbitrary scope. I think it needs to be added to a function, variable or class definition. But I added a `PARTHENON_WARN`.

pgrete commented on src/outputs/output_utils.hpp:
```diff
@@ -200,6 +203,59 @@ struct AllSwarmInfo {
+  const auto &Nt = pvar->GetDim(6);
+  const auto &Nu = pvar->GetDim(5);
+  const auto &Nv = pvar->GetDim(4);
+  const IndexDomain theDomain = (do_ghosts ? IndexDomain::entire : IndexDomain::interior);
```
Nitpicking style: `the_domain` or `domain`?

Yurlungur replied:
:+1: fixed

pgrete commented on src/outputs/output_utils.hpp:
```diff
@@ -200,6 +203,59 @@ struct AllSwarmInfo {
+template <typename Data_t, typename idx_t, typename Function_t>
+void PackOrUnpackVar(MeshBlock *pmb, Variable<Real> *pvar, bool do_ghosts, idx_t &idx,
+                     std::vector<Data_t> &data, Function_t f) {
+  auto v_h = pvar->data.GetHostMirrorAndCopy();
```
Just double checking (as host memory might actually be a constrained in some setups): Calling `GetHostMirrorAndCopy` from a `GetHostMirror` variable does *not* duplicate the memory allocation, correct?

Yurlungur replied:
I'm not sure I understand the question---you mean in the case where the memory space and host space are the same right? `GetHostMirrorAndCopy` is a thin wrapper (with our data structures) around `Kokkos::create_mirror_view_and_copy`, which only allocates new memory if the source and target memory spaces are different.

Yurlungur replied:
https://kokkos.github.io/kokkos-core-wiki/API/core/view/create_mirror.html?highlight=create_mirror_view_and_copy#_CPPv4I0DpE27create_mirror_view_and_copy14ImplMirrorTypeRKN4Impl12ViewCtorPropIDp12ViewCtorArgsEERK8ViewType

pgrete replied:
That's what I thought, which is why I was double checking (as currently there's another `GetHostMirror` right before calling this function is being called).

pgrete commented on src/outputs/parthenon_hdf5.cpp:
```diff
@@ -718,66 +595,128 @@ std::string PHDF5Output::GenerateFilename_(ParameterInput *pin, SimTime *tm,
+
+  // write Xmin[ndim] for blocks
+  {
+    // JMM: These arrays chould be shared, but I think this is clearer
+    // as to what's going on.
```
:+1: I stumbled across the original reuse pattern in the past...

pgrete commented on src/outputs/outputs.cpp:
```diff
@@ -263,6 +263,7 @@ Outputs::Outputs(Mesh *pm, ParameterInput *pin, SimTime *tm) {
           num_rst_outputs++;
         }
 #ifdef ENABLE_HDF5
+        op.write_xdmf = pin->GetOrAddBoolean(op.block_name, "write_xdmf", true);
```
Add to the doc that his param now exists?

Yurlungur replied:
:+1: done

pgrete commented on src/outputs/output_utils.cpp:
```diff
@@ -100,6 +100,71 @@ AllSwarmInfo::AllSwarmInfo(BlockList_t &block_list,
 
+// Tools that can be shared accross Output types
+
+std::vector<Real> ComputeXminBlocks(Mesh *pm) {
+  return FlattenBlockInfo<Real>(pm, 3,
+                                [=](MeshBlock *pmb, std::vector<Real> &data, int &i) {
+                                  auto xmin = pmb->coords.GetXmin();
+                                  data[i++] = xmin[0];
+                                  if (pm->ndim > 1) {
+                                    data[i++] = xmin[1];
+                                  }
+                                  if (pm->ndim > 2) {
+                                    data[i++] = xmin[2];
+                                  }
+                                });
```
Is there a potential danger for misuse by accident here for outputs with ndim != 3 because the returned vector's size will always be `3*num_blocks_local` while there might only be valid data in one or two thirds.

Yurlungur replied:
Hmm... yes there probably is. I believe this refactor maintains the behavior of the original, but let me think about edge cases here... Maybe the original code was a little dangerous.

Yurlungur replied:
I think there's no danger in changing `3` to `ndim` here so I'm just going to make that change.

pgrete reviewed (commented):
Overall I like these changes. Definitely make the easier to read and digest and allows for more reuse!

jonahm-LANL committed:
pgrete comments 1

jonahm-LANL committed:
3->ndim

jonahm-LANL committed:
changelog

jonahm-LANL committed:
CC

Yurlungur commented:
@bprather since @pgrete  wants to re-use this for his I/O stuff, let's get this first pass in now. Do you mind going through and reviewing?

@pgrete I *think* I've addressed all your comments. Please take another look.

Yurlungur commented on src/outputs/output_utils.hpp:
```diff
@@ -200,6 +203,59 @@ struct AllSwarmInfo {
+}
+
+template <typename Data_t, typename idx_t, typename Function_t>
+void PackOrUnpackVar(MeshBlock *pmb, Variable<Real> *pvar, bool do_ghosts, idx_t &idx,
```
Worth noting, @pgrete this is the one function I am likely to change down the road, if you wanted to use it. I plan to have this machinery return some shape information, and also permute it to support cells/edges/etc.

pgrete replied:
No worries.
I'll keep an eye on this.
Also I might later also modify this function in order to allow for coarsening of the data (so overall I'm very much in favor of having this abstraction/functionality).

jonahm-LANL committed:
deprecation

pgrete reviewed (approved):
Thanks for the updates. LGTM!

c-prather reviewed (approved):
LGTM this is a great start making this code more manageable.

Be warned I may use `PackOrUnpackVar` downstream, it's too convenient

pgrete commented:
I'm not sure why the test failed. I retriggered the test to see if it's reproducible

Yurlungur commented:
> I'm not sure why the test failed. I retriggered the test to see if it's reproducible

Seems to reproduce. I'll see if I can reproduce on my machine. Looks to be the restart test. Could have sworn I had this working on my machine, but... must have broken something in a recent commit.

Yurlungur commented:
> > I'm not sure why the test failed. I retriggered the test to see if it's reproducible
> 
> Seems to reproduce. I'll see if I can reproduce on my machine. Looks to be the restart test. Could have sworn I had this working on my machine, but... must have broken something in a recent commit.

Hmm ok I'm confused, because tests pass on my machine. I'll try to spin up a GPU machine and see if the failure is Cuda specific.

jonahm-LANL committed:
fix restart bug.

jonahm-LANL committed:
merge develop

jonahm-LANL committed:
formatting

Yurlungur commented:
CI failure located and fixed I believe. A silly bug in my pack or unpack variable
